fix: destination secret fields

saikumarrs opened the pull request:
## What are the changes introduced in this PR?

- Updated definitions of destinations that are in-violation of secrets configuration based on the following rules:
  - A field cannot be considered a secret if it is needed by at least one device mode integration as it'll be automatically exposed by the source configuration endpoint.
  - If a field is truly a secret, then it should not be part of `includeKeys`.
    - Userpilot, Refiner and AppsFlyer destinations.
- Updated validator to add the above custom rules for destination definition validation to prevent future violations.

## What is the related Linear task?

https://linear.app/rudderstack/issue/INT-4282/review-all-the-integration-definition-configurations-for-secret-fields

## Please explain the objectives of your changes below

DO NOT expose secret keys in source configuration response (via includeKeys).

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

Improved the validator to add custom rules beyond the JSON schema validation for destination definition configuration file.

### Any new dependencies introduced with this change?

No

### Any new checks got introduced or modified in test suites. Please explain the changes.

Destination definitions are now thoroughly validated to not expose secrets.

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] I have executed schemaGenerator tests and updated schema if needed

- [x] Are sensitive fields marked as secret in definition config?

- [x] My test cases and placeholders use only masked/sample values for sensitive fields

- [ ] Is the PR limited to 10 file changes & one task?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Added an npm "clean" script and bumped the package version; updated CI workflow action pins.

* **Configuration**
  * Updated many destination/source configs: adjusted which keys are included or treated as secrets; added deprecation and "new" flags and creation controls for deprecated items.

* **Features**
  * New post-schema validation enforcing include/exclude and device/hybrid-mode rules.

* **Tests**
  * Added comprehensive tests covering the new destination validation rules.

* **Documentation**
  * Clarified schema descriptions and UI-related comments for source/destination fields.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

saikumarrs committed:
chore: add clean script

saikumarrs committed:
chore: fix secrets config in destinations

saikumarrs committed:
fix: more desinations

saikumarrs committed:
chore: remove unwanted files

saikumarrs committed:
feat: add more validation rules

saikumarrs committed:
fix: handle undefined

saikumarrs committed:
fix: lint issues

saikumarrs committed:
fix: code coverage issues

saikumarrs committed:
fix: validation rules

saikumarrs committed:
fix: revert includeKeys and excludeKeys changes

saikumarrs committed:
chore: fix formatting

saikumarrs committed:
chore: fix lint issues

saikumarrs committed:
chore: revert unnecessary changes

saikumarrs committed:
chore: comment unused rules

saikumarrs committed:
chore: remove duplicate test case

saikumarrs commented on src/validator/index.ts:
```diff
@@ -16,6 +16,142 @@ const ajv = new Ajv({
+// Destination definition validation rules
+const destinationDefinitionRules: ValidationRule[] = [
+  {
+    description: 'Secret keys must not be in includeKeys unless also in excludeKeys',
```
The implementation is a bit weird in CBE.
https://github.com/rudderlabs/rudder-config-backend/blob/61c517e76ffe485f597c3510cf084dd164167bf0/src/modules/source-config/services/sourceConfig.service.ts#L99-L116
Seems like a bug.

Any field you can set with includeKeys, can be unset with excludeKeys. We might as well not specify that field in the includeKeys in the first place!
As a matter of fact, no destination configures excludeKeys array today.

The logic doesn't make sense at all. We need to fix the logic.

The ideal implementation should be something like this:
- If no arrays are configured, send all the fields.
- If either of the arrays are configured, filter the fields accordingly.
- Both arrays configured is an error situation. We may have to ignore excludeKeys.

For now, I've aligned the rules to match CBE implementation. Once we clean up there, we can fix it here.
WDYT?

devops-github-rudderstack commented:
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

saikumarrs committed:
chore: revert includeKeys changes (moved to separate branch)

Reverts includeKeys cleanup for 26 destinations and db-config-schema.json
changes to keep this branch focused on secretKeys cleanup only.

These changes are tracked in:
https://github.com/rudderlabs/rudder-integrations-config/pull/2367
